File: examples/LMemLoopback/client/cpp/AdvancedStatic/LMemLoopbackClient.cpp
/**
 * Copyright 2016 Maxeler Technologies
 *
 * Document: MaxCompiler Tutorial (maxcompiler-tutorial.pdf)
 * Chapter: 13      Example: 2      Name: LMem Loopback
 * MaxFile name: LMemLoopback
 * Summary:
 *        Adds two LMem input streams and writes the result to LMem.
 */
#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>

#include <stdlib.h>

#include <vector>

#include "../gen-cpp/LMemLoopbackService.h"

using std::cout;
using std::endl;

using apache::thrift::TException;
using apache::thrift::protocol::TBinaryProtocol;
using apache::thrift::protocol::TProtocol;
using apache::thrift::transport::TBufferedTransport;
using apache::thrift::transport::TSocket;
using apache::thrift::transport::TTransport;

using ::com::maxeler::LMemLoopback::LMemLoopbackServiceClient;
using ::com::maxeler::LMemLoopback::LMemLoopback_actions_t_struct;
using ::com::maxeler::LMemLoopback::LMemLoopback_readLMem_actions_t_struct;
using ::com::maxeler::LMemLoopback::LMemLoopback_writeLMem_actions_t_struct;
using ::com::maxeler::LMemLoopback::remote_ptr;

/**
 * Calculates how much time pasted from strartTime.
 *
 * @param startTime  Starting time
 *
 * @return           Time difference in seconds
 */
double calcTime(struct timeval startTime) {
  struct timeval endTime;

  gettimeofday(&endTime, NULL);

  double seconds  = endTime.tv_sec  - startTime.tv_sec;
  double useconds = endTime.tv_usec - startTime.tv_usec;

  return (((seconds) * 1000 + useconds / 1000.0) + 0.5) / 1000;
}

/**
 * Checks if LMemLoopbackDfe and LMemLoopbackCpu are the same.
 *
 * @param size       Size
 * @param outDataDfe Out data from DFE
 * @param outDataCpu Out data from CPU
 */
void check(std::vector<int32_t> outDataDFE,
           std::vector<int32_t> outDataCPU, int size) {
  for (int i = 0; i < size; i++) {
    if (outDataDFE[i] != outDataCPU[i]) {
      fprintf(stderr, "Output data @ %d = %d (expected %d)\n",
              i, outDataDFE[i], outDataCPU[i]);

      cout << "Test failed." << endl;
      exit(-1);
    }
  }

  cout << "Test passed!" << endl;
}

/**
 * LMemLoopback on CPU. outData = inA + inB
 *
 * @param size    Size
 * @param inA     In A
 * @param inB     In B
 *
 * @return        Out data
 */
std::vector<int32_t> LMemLoopbackCPU(int size, std::vector<int32_t> inA,
                                     std::vector<int32_t> inB) {
  std::vector<int32_t> outDataCPU(size);

  for (int i = 0; i < size; i++) {
    outDataCPU[i] = inA[i] + inB[i];
  }

  return outDataCPU;
}

/**
 * LMemLoopback on DFE. outData = inA + inB
 *
 * @param size    Size
 * @param inA     In A
 * @param inB     In B
 *
 * @return        Out data
 */
std::vector<int32_t> LMemLoopbackDFE(int size, std::vector<int32_t> inA,
                                     std::vector<int32_t> inB) {
  int sizeBytes = size * sizeof(int32_t);
  std::vector<int32_t> outData(size);

  struct timeval startTime;

  gettimeofday(&startTime, NULL);

  // Make socket
  boost::shared_ptr<TTransport> socket(new TSocket("localhost", 9090));

  // Buffering is critical. Raw sockets are very slow
  boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));

  // Wrap in a protocol
  boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));

  // Create a client to use the protocol encoder
  LMemLoopbackServiceClient client(protocol);

  cout << "Creating a client:\t\t\t\t";
  cout << calcTime(startTime) << "s" << endl;

  try {
    // Connect!
    gettimeofday(&startTime, NULL);
    transport->open();
    cout << "Opening connection:\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Initialize maxfile
    gettimeofday(&startTime, NULL);
    remote_ptr maxfile = client.LMemLoopback_init();
    cout << "Initializing maxfile:\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Load DFE
    gettimeofday(&startTime, NULL);
    remote_ptr engine = client.max_load(maxfile, "*");
    cout << "Loading DFE:\t\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Allocate and send input streams to server
    gettimeofday(&startTime, NULL);

    remote_ptr address_inA = client.malloc_int32_t(size);
    client.send_data_int32_t(address_inA, inA);

    remote_ptr address_inB = client.malloc_int32_t(size);
    client.send_data_int32_t(address_inB, inB);

    cout << "Sending input data:\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Allocate memory for output stream on server
    gettimeofday(&startTime, NULL);
    remote_ptr address_outData = client.malloc_int32_t(size);
    cout << "Allocating memory for output stream on server:\t";
    cout << calcTime(startTime) << "s" << endl;

    // Write to LMem
    gettimeofday(&startTime, NULL);

    LMemLoopback_writeLMem_actions_t_struct actions_lmem_write;

    actions_lmem_write.__set_param_address(0);
    actions_lmem_write.__set_param_nbytes(sizeBytes);
    actions_lmem_write.__set_instream_cpu_to_lmem(address_inA);

    remote_ptr address_actions_lmem_write =
        client.send_LMemLoopback_writeLMem_actions_t(actions_lmem_write);
    client.LMemLoopback_writeLMem_run(engine, address_actions_lmem_write);

    actions_lmem_write.__set_param_address(sizeBytes);
    actions_lmem_write.__set_param_nbytes(sizeBytes);
    actions_lmem_write.__set_instream_cpu_to_lmem(address_inB);

    address_actions_lmem_write =
        client.send_LMemLoopback_writeLMem_actions_t(actions_lmem_write);
    client.LMemLoopback_writeLMem_run(engine, address_actions_lmem_write);

    cout << "Writing to LMem:\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Action default
    gettimeofday(&startTime, NULL);

    LMemLoopback_actions_t_struct actions;

    actions.__set_param_N(size);

    remote_ptr address_actions = client.send_LMemLoopback_actions_t(actions);
    client.LMemLoopback_run(engine, address_actions);

    cout << "LMemLoopback time:\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Reading from LMem
    gettimeofday(&startTime, NULL);

    LMemLoopback_readLMem_actions_t_struct actions_lmem_read;

    actions_lmem_read.__set_param_address(2 * sizeBytes);
    actions_lmem_read.__set_param_nbytes(sizeBytes);
    actions_lmem_read. __set_outstream_lmem_to_cpu(address_outData);

    remote_ptr address_actions_lmem_read =
        client.send_LMemLoopback_readLMem_actions_t(actions_lmem_read);
    client.LMemLoopback_readLMem_run(engine, address_actions_lmem_read);

    cout << "Reading from LMem:\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Unload DFE
    gettimeofday(&startTime, NULL);
    client.max_unload(engine);
    cout << "Unloading DFE:\t\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Get output stream from server
    gettimeofday(&startTime, NULL);
    client.receive_data_int32_t(outData, address_outData, size);
    cout << "Getting output stream:\t(size = " << size * 32 << " bit)\t";
    cout << calcTime(startTime) << "s" << endl;

    // Free allocated memory for streams on server
    gettimeofday(&startTime, NULL);
    client.free(address_inA);
    client.free(address_inB);
    client.free(address_outData);
    client.free(address_actions);
    client.free(address_actions_lmem_read);
    client.free(address_actions_lmem_write);
    cout << "Freeing allocated memory for streams on server:\t";
    cout << calcTime(startTime) << "s" << endl;

    // Free allocated maxfile data
    gettimeofday(&startTime, NULL);
    client.LMemLoopback_free();
    cout << "Freeing allocated maxfile data:\t\t\t";
    cout << calcTime(startTime) << "s" << endl;

    // Close!
    gettimeofday(&startTime, NULL);
    transport->close();
    cout << "Closing connection:\t\t\t\t";
    cout << calcTime(startTime) << "s" << endl;
  } catch (TException& tx) {
    cout << "ERROR: " << tx.what() << endl;
    exit(-1);
  }

  return outData;
}

/**
 * Calculates LMemLoopbackDFE and LMemLoopbackCPU and
 * checks if they return the same value.
 */
int main() {
  struct timeval startTime;

  // Input
  gettimeofday(&startTime, NULL);

  const int size = 384;
  std::vector<int32_t> inA(size);
  std::vector<int32_t> inB(size);

  for (int i = 0; i < size; i++) {
    inA[i] = i;
    inB[i] = size - i;
  }

  cout << "Generating input data:\t\t\t\t";
  cout << calcTime(startTime) << "s" << endl;

  // DFE Output
  gettimeofday(&startTime, NULL);
  std::vector<int32_t> outDataDFE = LMemLoopbackDFE(size, inA, inB);
  cout << "DFE LMemLoopback total time:\t\t\t";
  cout << calcTime(startTime) << "s" << endl;

  // CPU Output
  gettimeofday(&startTime, NULL);
  std::vector<int32_t> outDataCPU = LMemLoopbackCPU(size, inA, inB);
  cout << "CPU LMemLoopback total time:\t\t\t";
  cout << calcTime(startTime) << "s" << endl;

  // Checking results
  check(outDataCPU, outDataDFE, size);

  return 0;
}

File: examples/LMemLoopback/client/py/AdvancedStatic/LMemLoopbackClient.py
#!/usr/bin/env python
"""LMemLoopback example"""

import sys
import time
sys.path.append('../gen-py')

from thrift import Thrift
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol

from com.maxeler.LMemLoopback import LMemLoopbackService
from com.maxeler.LMemLoopback.ttypes import (
    LMemLoopback_readLMem_actions_t_struct,
    LMemLoopback_writeLMem_actions_t_struct,
    LMemLoopback_actions_t_struct)

def check(data_out_dfe, data_out_cpu, size):
    """Check if data_out_dfe and data_out_cpu match."""
    status = 0
    for i in range(size):
        if data_out_dfe[i] != data_out_cpu[i]:
            print str(data_out_dfe[i]) + " != " + str(data_out_cpu[i])
            status = status + 1
    return status

def lmem_loopback_cpu(size, in_a, in_b):
    """LMemLoopback CPU implementation."""
    return [in_a[i] + in_b[i] for i in range(size)]

def lmem_loopback_dfe(size, in_a, in_b):
    """LMemLoopback DFE implementation."""
    try:
        start_time = time.time()
        # Make socket
        socket = TSocket.TSocket('localhost', 9090)

        # _buffering is critical. Raw sockets are very slow
        transport = TTransport.TBufferedTransport(socket)

        # Wrap in a protocol
        protocol = TBinaryProtocol.TBinaryProtocol(transport)

        # Create a client to use the protocol encoder
        client = LMemLoopbackService.Client(protocol)
        print ('Creating a client:\t\t\t\t%.5lfs' %
       	       (time.time() - start_time))

        # Connect!
        start_time = time.time()
        transport.open()
        print ('Opening connection:\t\t\t\t%.5lfs' %
               (time.time() - start_time))

        size_bytes = size * 4

        # Initialize maxfile
        start_time = time.time()
        max_file = client.LMemLoopback_init()
        print ('Initializing maxfile:\t\t\t\t%.5lfs' %
               (time.time() - start_time))

        # Load DFE
        start_time = time.time()
        max_engine = client.max_load(max_file, '*')
        print ('Loading DFE:\t\t\t\t\t%.5lfs' %
               (time.time() - start_time))

        # Allocate and send input streams to server
        start_time = time.time()
        address_in_a = client.malloc_int32_t(size)
        client.send_data_int32_t(address_in_a, in_a)

        address_in_b = client.malloc_int32_t(size)
        client.send_data_int32_t(address_in_b, in_b)
        print ('Sending input data:\t\t\t\t%.5lfs' %
               (time.time() - start_time))

        # Allocate memory for output stream on server
        start_time = time.time()
        address_out_data = client.malloc_int32_t(size)
        print ('Allocating memory for output stream on server:\t%.5lfs'%
               (time.time() - start_time))

        # Write to LMem
        start_time = time.time()
        actions_write_lmem = (
            LMemLoopback_writeLMem_actions_t_struct(0, size_bytes,
                                                    address_in_a))
        address_actions_write_lmem = (
            client.send_LMemLoopback_writeLMem_actions_t(actions_write_lmem))
        client.LMemLoopback_writeLMem_run(max_engine,
                                          address_actions_write_lmem)

        actions_write_lmem = (
            LMemLoopback_writeLMem_actions_t_struct(size_bytes, size_bytes,
                                                    address_in_b))
        address_actions_write_lmem = (
            client.send_LMemLoopback_writeLMem_actions_t(actions_write_lmem))
        client.LMemLoopback_writeLMem_run(max_engine,
                                          address_actions_write_lmem)
        print ('Writing to LMem:\t\t\t\t%.5lfs'%
               (time.time() - start_time))

	# Action default
        start_time = time.time()
        actions = LMemLoopback_actions_t_struct(size)
        address_actions = client.send_LMemLoopback_actions_t(actions)
        client.LMemLoopback_run(max_engine, address_actions)
        print ('LMemLoopback time:\t\t\t\t%.5lfs'%
               (time.time() - start_time))

        # Read from LMem
        start_time = time.time()
        actions_read_lmem = (
            LMemLoopback_readLMem_actions_t_struct(2 * size_bytes, size_bytes,
                                                   address_out_data))
        address_actions_read_lmem = (
            client.send_LMemLoopback_readLMem_actions_t(actions_read_lmem))
        client.LMemLoopback_readLMem_run(max_engine, address_actions_read_lmem)
        print ('Reading from LMem:\t\t\t\t%.5lfs'%
               (time.time() - start_time))

        # Unload DFE
        start_time = time.time()
        client.max_unload(max_engine)
        print ('Unloading DFE:\t\t\t\t\t%.5lfs' %
               (time.time() - start_time))

        # Get output stream from server
        start_time = time.time()
        out_data = client.receive_data_int32_t(address_out_data, size)
        print ('Getting output stream:\t(size = %d bit)\t%.5lfs' %
               ((size * 32), (time.time() - start_time)))

        # Free allocated memory for streams on server
        start_time = time.time()
        client.free(address_in_a)
        client.free(address_in_b)
        client.free(address_out_data)
        client.free(address_actions_write_lmem)
        client.free(address_actions)
        client.free(address_actions_read_lmem)
        print ('Freeing allocated memory for streams on server:\t%.5lfs' %
               (time.time() - start_time))

        # Free allocated maxfile data
        start_time = time.time()
        client.LMemLoopback_free()
        print ('Freeing allocated maxfile data:\t\t\t%.5lfs' %
               (time.time() - start_time))

        # Close!
        start_time = time.time()
        transport.close()
        print ('Closing connection:\t\t\t\t%.5lfs' %
               (time.time() - start_time))


    except Thrift.TException, thrift_exceptiion:
        print '%s' % (thrift_exceptiion.message)
        sys.exit(-1)

    return out_data

def test():
    """
    Calls LMemLoopbackDFE and LMemLoopbackCPU
    and checks if they return the same result.
    """
    # Input
    start_time = time.time()
    size = 384

    in_a = [i for i in range(size)]
    in_b = [size - i for i in range(size)]
    print ('Generating input data:\t\t\t\t%.5lfs' %
           (time.time() - start_time))

    # DFE Output
    start_time = time.time()
    data_out_dfe = lmem_loopback_dfe(size, in_a, in_b)
    print ('DFE LMemLoopback total time:\t\t\t%.5lfs' %
           (time.time() - start_time))

    # CPU Output
    start_time = time.time()
    data_out_cpu = lmem_loopback_cpu(size, in_a, in_b)
    print ('CPU LMemLoopback total time:\t\t\t%.5lfs' %
           (time.time() - start_time))

    # Checking results
    start_time = time.time()
    status = check(data_out_dfe, data_out_cpu, size)
    print ('Checking results:\t\t\t\t%.5lfs' %
           (time.time() - start_time))

    if status == 0:
        print "Test passed!"
    else:
        print "Test failed %.5d times!" % (status)
        sys.exit(-1)

if __name__ == '__main__':
    test()

